File: swift/timesync.swift
#!/usr/bin/env swift
//
// timesync.swift - Minimal SNTP client (RFC 5905 subset)
//
// SPDX-License-Identifier: MIT
// Copyright (c) 2025 tsupplis
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//
// Swift port of the C implementation
//
// Build:
//   swiftc -O -o timesync timesync.swift
//
// Usage:
//   ./timesync                    # query pool.ntp.org
//   ./timesync -t 1500 -r 2 -v time.google.com

import Foundation
#if canImport(Darwin)
import Darwin
#elseif canImport(Glibc)
import Glibc
#endif

// MARK: - Constants

let NTP_PORT: UInt16 = 123
let NTP_PACKET_SIZE = 48
let NTP_UNIX_EPOCH_DIFF: UInt64 = 2208988800
let DEFAULT_SERVER = "pool.ntp.org"
let DEFAULT_TIMEOUT_MS = 2000
let DEFAULT_RETRIES = 3

// MARK: - Config

struct Config {
    var server: String = DEFAULT_SERVER
    var timeoutMs: Int = DEFAULT_TIMEOUT_MS
    var retries: Int = DEFAULT_RETRIES
    var verbose: Bool = false
    var testOnly: Bool = false
    var useSyslog: Bool = false
}

// MARK: - Logging

/// Log to stderr with timestamp — matches C's stderr_log format: "YYYY-MM-DD HH:MM:SS <message>"
func stderrLog(_ message: String) {
    var now = time(nil)
    var tm = tm()
    localtime_r(&now, &tm)
    var buf = [CChar](repeating: 0, count: 32)
    strftime(&buf, buf.count, "%Y-%m-%d %H:%M:%S", &tm)
    let timestamp = String(cString: buf)
    fputs("\(timestamp) \(message)\n", stderr)
}

/// Write a message to syslog at the given priority.
/// Uses vsyslog+withVaList to work around Swift's restriction on variadic C functions.
func syslogWrite(_ priority: Int32, _ message: String) {
    message.withCString { msg in
        withVaList([msg]) { args in vsyslog(priority, "%s", args) }
    }
}

// MARK: - Usage

func showUsage() {
    let prog = CommandLine.arguments[0]
    fputs("Usage: \(prog) [-t timeout_ms] [-r retries] [-n] [-v] [-s] [-h] [ntp server]\n", stderr)
    fputs("  server       NTP server to query (default: pool.ntp.org)\n", stderr)
    fputs("  -t timeout   Timeout in ms (default: 2000)\n", stderr)
    fputs("  -r retries   Number of retries (default: 3)\n", stderr)
    fputs("  -n           Test mode (no system time adjustment)\n", stderr)
    fputs("  -v           Verbose output\n", stderr)
    fputs("  -s           Enable syslog logging\n", stderr)
    fputs("  -h           Show this help message\n", stderr)
    exit(0)
}

// MARK: - Argument parsing

func parseArgs() -> Config {
    var config = Config()
    let args = CommandLine.arguments
    var i = 1

    while i < args.count {
        let arg = args[i]
        switch arg {
        case "-h":
            showUsage()
        case "-t":
            i += 1
            if i < args.count, let v = Int(args[i]) {
                // invalid/0 resets to default, >6000 clamps — matches C
                if v <= 0 { config.timeoutMs = DEFAULT_TIMEOUT_MS }
                else if v > 6000 { config.timeoutMs = 6000 }
                else { config.timeoutMs = v }
            }
        case "-r":
            i += 1
            if i < args.count, let v = Int(args[i]) {
                // invalid/0 resets to default, >10 clamps — matches C
                if v <= 0 { config.retries = DEFAULT_RETRIES }
                else if v > 10 { config.retries = 10 }
                else { config.retries = v }
            }
        case "-n": config.testOnly = true
        case "-v": config.verbose = true
        case "-s": config.useSyslog = true
        default:
            if !arg.hasPrefix("-") { config.server = arg }
            // unknown flags silently ignored like C
        }
        i += 1
    }

    if config.testOnly { config.useSyslog = false }
    return config
}

// MARK: - Time helpers

/// Current time in ms since Unix epoch — mirrors C's gettimeofday().
func getTimeMs() -> Int64 {
    var tv = timeval()
    gettimeofday(&tv, nil)
    return Int64(tv.tv_sec) * 1000 + Int64(tv.tv_usec) / 1000
}

/// Format ms-since-epoch as local ISO string with ms suffix — matches C's %Y-%m-%dT%H:%M:%S%z.
func formatTime(_ ms: Int64) -> String {
    var secs = time_t(ms / 1000)
    var tm = tm()
    localtime_r(&secs, &tm)
    var buf = [CChar](repeating: 0, count: 64)
    strftime(&buf, buf.count, "%Y-%m-%dT%H:%M:%S%z", &tm)
    let msSuffix = String(format: "%03d", Int(ms % 1000))
    return String(cString: buf) + "." + msSuffix
}

// MARK: - NTP protocol

/// Build a 48-byte NTP request packet — LI=0, VN=4, Mode=3 → 0x23, matching C.
func buildNtpRequest() -> [UInt8] {
    var packet = [UInt8](repeating: 0, count: NTP_PACKET_SIZE)
    packet[0] = 0x23
    return packet
}

/// Convert 8-byte big-endian NTP timestamp at buf[offset] to Unix ms.
func ntpTsToUnixMs(_ buf: [UInt8], offset: Int) -> Int64? {
    guard offset + 8 <= buf.count else { return nil }
    let sec = UInt64(buf[offset]) << 24 | UInt64(buf[offset+1]) << 16
            | UInt64(buf[offset+2]) << 8  | UInt64(buf[offset+3])
    let frac = UInt64(buf[offset+4]) << 24 | UInt64(buf[offset+5]) << 16
             | UInt64(buf[offset+6]) << 8  | UInt64(buf[offset+7])
    guard sec >= NTP_UNIX_EPOCH_DIFF else { return nil }
    let usec = (frac * 1_000_000) >> 32
    let unixSec = sec - NTP_UNIX_EPOCH_DIFF
    return Int64(unixSec * 1000 + usec / 1000)
}

// MARK: - NTP query

struct NtpResponse {
    let localBeforeMs: Int64
    let remoteMs: Int64
    let localAfterMs: Int64
    let serverAddr: String
}

/// Single NTP query attempt. Returns NtpResponse on success, nil on any failure.
func doNtpQuery(server: String, timeoutMs: Int) -> NtpResponse? {
    // Resolve server — mirrors C's getaddrinfo(AF_UNSPEC, SOCK_DGRAM)
    var hints = addrinfo()
    hints.ai_family = AF_UNSPEC
    hints.ai_socktype = SOCK_DGRAM
    var res: UnsafeMutablePointer<addrinfo>? = nil
    guard getaddrinfo(server, String(NTP_PORT), &hints, &res) == 0, let addrList = res else { return nil }
    defer { freeaddrinfo(addrList) }

    var rp: UnsafeMutablePointer<addrinfo>? = addrList
    while let node = rp {
        defer { rp = node.pointee.ai_next }

        let sockfd = socket(node.pointee.ai_family, node.pointee.ai_socktype, node.pointee.ai_protocol)
        guard sockfd >= 0 else { continue }
        defer { close(sockfd) }

        // Set receive timeout — mirrors C's setsockopt SO_RCVTIMEO
        var tv = timeval()
        tv.tv_sec = timeoutMs / 1000
        tv.tv_usec = Int32((timeoutMs % 1000) * 1000)
        if setsockopt(sockfd, SOL_SOCKET, SO_RCVTIMEO, &tv, socklen_t(MemoryLayout<timeval>.size)) < 0 {
            stderrLog("WARNING setsockopt SO_RCVTIMEO failed: \(String(cString: strerror(errno)))")
            continue
        }

        let packet = buildNtpRequest()
        let localBefore = getTimeMs()

        // Send
        let sent = packet.withUnsafeBytes { bytes in
            sendto(sockfd, bytes.baseAddress, NTP_PACKET_SIZE, 0,
                   node.pointee.ai_addr, node.pointee.ai_addrlen)
        }
        guard sent == NTP_PACKET_SIZE else {
            stderrLog("WARNING Failed to send NTP request")
            continue
        }

        // Receive
        var buf = [UInt8](repeating: 0, count: NTP_PACKET_SIZE)
        var srcAddr = sockaddr_storage()
        var srcLen = socklen_t(MemoryLayout<sockaddr_storage>.size)
        let recvd = buf.withUnsafeMutableBytes { bytes in
            withUnsafeMutablePointer(to: &srcAddr) { srcPtr in
                srcPtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockPtr in
                    recvfrom(sockfd, bytes.baseAddress, NTP_PACKET_SIZE, 0, sockPtr, &srcLen)
                }
            }
        }
        let localAfter = getTimeMs()

        guard recvd >= NTP_PACKET_SIZE else { continue }

        // Validate mode = 4 (server)
        let mode = buf[0] & 0x07
        guard mode == 4 else {
            stderrLog("WARNING Invalid mode in NTP response: \(mode)")
            continue
        }
        // Validate stratum != 0
        let stratum = buf[1]
        guard stratum != 0 else {
            stderrLog("WARNING Invalid stratum in NTP response: \(stratum)")
            continue
        }
        // Validate version 1-4
        let version = (buf[0] >> 3) & 0x07
        guard (1...4).contains(version) else {
            stderrLog("WARNING Invalid version in NTP response: \(version)")
            continue
        }
        // Extract transmit timestamp at bytes 40-47
        guard let remoteMs = ntpTsToUnixMs(buf, offset: 40) else {
            stderrLog("WARNING Invalid transmit timestamp in NTP response")
            continue
        }

        // Compose server address string from the source address
        var ipStr = [CChar](repeating: 0, count: Int(INET6_ADDRSTRLEN))
        withUnsafePointer(to: &srcAddr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sa in
                switch Int32(sa.pointee.sa_family) {
                case AF_INET:
                    ptr.withMemoryRebound(to: sockaddr_in.self, capacity: 1) {
                        var addr = $0.pointee.sin_addr
                        inet_ntop(AF_INET, &addr, &ipStr, socklen_t(INET6_ADDRSTRLEN))
                    }
                case AF_INET6:
                    ptr.withMemoryRebound(to: sockaddr_in6.self, capacity: 1) {
                        var addr = $0.pointee.sin6_addr
                        inet_ntop(AF_INET6, &addr, &ipStr, socklen_t(INET6_ADDRSTRLEN))
                    }
                default: break
                }
            }
        }
        let serverAddrStr = String(cString: ipStr)

        return NtpResponse(localBeforeMs: localBefore, remoteMs: remoteMs,
                           localAfterMs: localAfter, serverAddr: serverAddrStr)
    }
    return nil
}

// MARK: - Set system time

func setSystemTime(_ ms: Int64) -> Bool {
    var tv = timeval()
    tv.tv_sec = Int(ms / 1000)
    tv.tv_usec = Int32((ms % 1000) * 1000)
    return settimeofday(&tv, nil) == 0
}

// MARK: - Main logic

func run(_ config: Config) -> Int32 {
    // Verbose debug fires before syslog is opened — matching C
    if config.verbose {
        stderrLog("DEBUG Using server: \(config.server)")
        stderrLog("DEBUG Timeout: \(config.timeoutMs) ms, Retries: \(config.retries), Syslog: \(config.useSyslog ? "on" : "off")")
    }

    if config.useSyslog {
        openlog("ntp_client", LOG_PID | LOG_CONS, LOG_USER)
        // Mirror C's atexit(closelog) — close syslog on process exit
        atexit { closelog() }
    }

    var response: NtpResponse? = nil
    for attempt in 0..<config.retries {
        if config.verbose {
            stderrLog("DEBUG Attempt (\(attempt + 1)) at NTP query on \(config.server) ...")
        }
        if let r = doNtpQuery(server: config.server, timeoutMs: config.timeoutMs) {
            response = r
            break
        }
        // Backoff fires on every failure including after last attempt — matches C
        usleep(200_000)
    }

    guard let resp = response else {
        stderrLog("ERROR Failed to contact NTP server \(config.server) after \(config.retries) attempts")
        if config.useSyslog {
            syslogWrite(LOG_ERR, "NTP query failed for \(config.server) after \(config.retries) attempts")
        }
        return 2
    }

    // Overflow guard on avg calculation — matches C's INT64_MAX check
    guard resp.localBeforeMs <= Int64.max - resp.localAfterMs else {
        stderrLog("ERROR Time averaging would overflow, invalid timestamps.")
        if config.useSyslog { syslogWrite(LOG_ERR, "Time averaging would overflow") }
        return 1
    }
    let avgLocalMs = (resp.localBeforeMs + resp.localAfterMs) / 2
    let offsetMs   = resp.remoteMs - avgLocalMs
    let roundtripMs = resp.localAfterMs - resp.localBeforeMs

    // Parse remote time unconditionally before verbose block — matches C's execution order
    let remoteTimeStr = formatTime(resp.remoteMs)
    var remoteSecs = time_t(resp.remoteMs / 1000)
    var remoteTm = tm()
    localtime_r(&remoteSecs, &remoteTm)
    let remoteYear = Int(remoteTm.tm_year) + 1900

    if config.verbose {
        stderrLog("DEBUG Server: \(config.server) (\(resp.serverAddr))")
        // local time: date from localBeforeMs, ms suffix from localAfterMs — matches C
        var localSecs = time_t(resp.localBeforeMs / 1000)
        var localTm = tm()
        localtime_r(&localSecs, &localTm)
        var localBuf = [CChar](repeating: 0, count: 64)
        strftime(&localBuf, localBuf.count, "%Y-%m-%dT%H:%M:%S%z", &localTm)
        let localMsSuffix = String(format: "%03d", Int(resp.localAfterMs % 1000))
        let localTimeStr = String(cString: localBuf) + "." + localMsSuffix
        stderrLog("DEBUG Local time: \(localTimeStr)")
        stderrLog("DEBUG Remote time: \(remoteTimeStr)")
        stderrLog("DEBUG Local before(ms): \(resp.localBeforeMs)")
        stderrLog("DEBUG Local after(ms): \(resp.localAfterMs)")
        stderrLog("DEBUG Estimated roundtrip(ms): \(roundtripMs)")
        stderrLog("DEBUG Estimated offset remote - local(ms): \(offsetMs)")
        if config.useSyslog {
            syslogWrite(LOG_INFO, "NTP server=\(config.server) addr=\(resp.serverAddr) offset_ms=\(offsetMs) rtt_ms=\(roundtripMs)")
        }
    }

    // Roundtrip sanity check
    guard roundtripMs >= 0 && roundtripMs <= 10000 else {
        stderrLog("ERROR Invalid roundtrip time: \(roundtripMs) ms")
        if config.useSyslog {
            syslogWrite(LOG_ERR, "Invalid suspiciously long roundtrip time: \(roundtripMs) ms")
        }
        return 1
    }

    // Delta < 500ms: skip adjustment
    if abs(offsetMs) > 0 && abs(offsetMs) < 500 {
        if config.verbose {
            stderrLog("INFO Delta < 500ms, not setting system time.")
            if config.useSyslog { syslogWrite(LOG_INFO, "Delta < 500ms, not setting system time") }
        }
        return 0
    }

    // Year range check
    guard remoteYear >= 2025 && remoteYear <= 2200 else {
        stderrLog("ERROR Remote year is out of valid range (2025-2200): \(remoteYear)")
        if config.useSyslog {
            syslogWrite(LOG_ERR, "Remote year is out of valid range (2025-2200): \(remoteYear)")
        }
        return 1
    }

    if config.testOnly { return 0 }

    // Root check
    guard getuid() == 0 else {
        stderrLog("WARNING Not root, not setting system time.")
        if config.useSyslog { syslogWrite(LOG_WARNING, "Not root, not setting system time") }
        return 0
    }

    // Overflow guard on remote_ms + half_rtt — matches C's INT64_MAX check
    let halfRtt = roundtripMs / 2
    guard resp.remoteMs <= Int64.max - halfRtt else {
        stderrLog("ERROR Time calculation would overflow, not adjusting system time.")
        if config.useSyslog { syslogWrite(LOG_ERR, "Time calculation would overflow") }
        return 1
    }
    let newTimeMs = resp.remoteMs + halfRtt

    let api = "settimeofday"
    if setSystemTime(newTimeMs) {
        stderrLog("INFO System time set using \(api) (\(remoteTimeStr))")
        if config.useSyslog { syslogWrite(LOG_INFO, "System time set using \(api) (\(remoteTimeStr))") }
        return 0
    } else {
        stderrLog("ERROR Failed to adjust system time with \(api): \(String(cString: strerror(errno)))")
        if config.useSyslog {
            syslogWrite(LOG_ERR, "Failed to adjust system time with \(api): \(String(cString: strerror(errno)))")
        }
        return 10
    }
}

// MARK: - Entry point

let config = parseArgs()
exit(run(config))

File: perl/timesync.pl
#!/usr/bin/env perl
#
# timesync - Minimal SNTP client (RFC 5905 subset)
#
# SPDX-License-Identifier: MIT
# Copyright (c) 2025 tsupplis
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

use strict;
use warnings;
use Socket;
use Time::HiRes qw(gettimeofday);
use Sys::Syslog qw(:standard :macros);
use POSIX qw(strftime);

# Constants
use constant NTP_PACKET_SIZE => 48;
# Number of seconds between 1900 (NTP epoch) and 1970 (Unix epoch)
use constant NTP_UNIX_EPOCH  => 2208988800;
use constant {
    NTP_PORT           => 123,
    DEFAULT_SERVER     => 'pool.ntp.org',
    DEFAULT_TIMEOUT_MS => 2000,
    DEFAULT_RETRIES    => 3,
};

# Configuration
my %config = (
    server      => DEFAULT_SERVER,
    timeout_ms  => DEFAULT_TIMEOUT_MS,
    retries     => DEFAULT_RETRIES,
    verbose     => 0,
    test_only   => 0,
    use_syslog  => 0,
);

# Logging function with time prefix (always to stderr)
sub stderr_log {
    my $msg = shift;
    my $ts = strftime('%Y-%m-%d %H:%M:%S', localtime);
    warn "$ts $msg\n";
}

sub log_syslog {
    return unless $config{use_syslog};
    my ($level, $msg) = @_;
    syslog($level, '%s', $msg);
}

sub show_usage {
    print STDERR "Usage: timesync [-t timeout_ms] [-r retries] [-n] [-v] [-s] [-h] [ntp server]\n";
    print STDERR "  server       NTP server to query (default: pool.ntp.org)\n";
    print STDERR "  -t timeout   Timeout in ms (default: 2000)\n";
    print STDERR "  -r retries   Number of retries (default: 3)\n";
    print STDERR "  -n           Test mode (no system time adjustment)\n";
    print STDERR "  -v           Verbose output\n";
    print STDERR "  -s           Enable syslog logging\n";
    print STDERR "  -h           Show this help message\n";
}

# Parse command line
while (@ARGV) {
    my $arg = shift @ARGV;
    if ($arg eq '-h') {
        show_usage();
        exit 0;
    } elsif ($arg eq '-t' && @ARGV) {
        $config{timeout_ms} = shift @ARGV;
        $config{timeout_ms} = 2000 if $config{timeout_ms} <= 0;
        $config{timeout_ms} = 6000 if $config{timeout_ms} > 6000;
    } elsif ($arg eq '-r' && @ARGV) {
        $config{retries} = shift @ARGV;
        $config{retries} = 3  if $config{retries} <= 0;
        $config{retries} = 10 if $config{retries} > 10;
    } elsif ($arg =~ /^-([nvsh]+)$/) {
        # Handle combined flags like -nv
        my $flags = $1;
        if ($flags =~ /h/) { show_usage(); exit 0; }
        $config{test_only}  = 1 if $flags =~ /n/;
        $config{verbose}    = 1 if $flags =~ /v/;
        $config{use_syslog} = 1 if $flags =~ /s/;
    } elsif ($arg !~ /^-/) {
        $config{server} = $arg;
    }
}

# Disable syslog in test mode
$config{use_syslog} = 0 if $config{test_only};

if ($config{verbose}) {
    stderr_log("DEBUG Using server: $config{server}");
    stderr_log(sprintf("DEBUG Timeout: %d ms, Retries: %d, Syslog: %s",
        $config{timeout_ms}, $config{retries},
        $config{use_syslog} ? 'on' : 'off'));
}

# Open syslog if needed
openlog('ntp_client', 'pid,cons', LOG_USER) if $config{use_syslog};

# Get time in milliseconds
sub get_time_ms {
    my ($sec, $usec) = gettimeofday();
    return $sec * 1000 + int($usec / 1000);
}

# Build NTP request packet (48 bytes)
# LI=0, VN=4, Mode=3 (client) -> 0b00100011 = 0x23
# rest are zero (we don't set transmit timestamp; some servers prefer it zero)
sub build_ntp_request {
    return pack('C', 0x23) . ("\0" x (NTP_PACKET_SIZE - 1));
}

# Read 64-bit NTP timestamp from buffer (bytes are big-endian).
# NTP timestamp: seconds (32 bits) + fractional (32 bits).
# We convert to milliseconds since Unix epoch.
sub ntp_to_unix_ms {
    my ($buf, $offset) = @_;

    my $sec  = unpack('N', substr($buf, $offset,     4));
    my $frac = unpack('N', substr($buf, $offset + 4, 4));

    return -1 if $sec < NTP_UNIX_EPOCH;

    my $unix_sec = $sec - NTP_UNIX_EPOCH;
    my $usec     = ($frac * 1_000_000) >> 32;  # fraction to microseconds

    return $unix_sec * 1000 + int($usec / 1000);
}

# Format time in ISO 8601 local time with timezone offset
sub format_time_local {
    my $ms  = shift;
    my $sec = int($ms / 1000);
    return strftime('%Y-%m-%dT%H:%M:%S%z', localtime($sec));
}

# Perform a single NTP query attempt; returns (0, results...) on success or (-1) on failure
sub do_ntp_query {
    my ($server, $timeout_ms) = @_;

    # Resolve hostname
    my @addrs = gethostbyname($server);
    return -1 unless @addrs;

    my $ip = inet_ntoa($addrs[4]);

    # Create UDP socket
    socket(my $sock, AF_INET, SOCK_DGRAM, getprotobyname('udp'))
        or return -1;

    # Set receive timeout
    my $timeout_tv = pack('L!L!', int($timeout_ms / 1000),
                          ($timeout_ms % 1000) * 1000);
    unless (setsockopt($sock, SOL_SOCKET, SO_RCVTIMEO, $timeout_tv)) {
        stderr_log("WARNING setsockopt SO_RCVTIMEO failed: $!");
        close($sock);
        return -1;
    }

    # Build and send NTP request
    my $packet = build_ntp_request();
    my $addr   = sockaddr_in(NTP_PORT, inet_aton($ip));

    my $local_before = get_time_ms();
    my $sent = send($sock, $packet, 0, $addr);
    if (!defined $sent || $sent != NTP_PACKET_SIZE) {
        stderr_log("WARNING Failed to send NTP request");
        close($sock);
        return -1;
    }

    # Receive response
    my $response;
    my $recv_addr = recv($sock, $response, NTP_PACKET_SIZE, 0);
    my $local_after = get_time_ms();

    close($sock);

    # Must have a full packet
    return -1 unless defined $recv_addr && length($response) >= NTP_PACKET_SIZE;

    # Validate mode = 4 (server)
    my $first_byte = unpack('C', substr($response, 0, 1));
    my $mode       = $first_byte & 0x07;
    if ($mode != 4) {
        stderr_log("WARNING Invalid mode in NTP response: $mode");
        return -1;
    }

    # Validate stratum != 0
    my $stratum = unpack('C', substr($response, 1, 1));
    if ($stratum == 0) {
        stderr_log("WARNING Invalid stratum in NTP response: $stratum");
        return -1;
    }

    # Validate version (1-4 valid)
    my $version = ($first_byte >> 3) & 0x07;
    if ($version < 1 || $version > 4) {
        stderr_log("WARNING Invalid version in NTP response: $version");
        return -1;
    }

    # Parse transmit timestamp (bytes 40-47)
    my $remote_ms = ntp_to_unix_ms($response, 40);
    if ($remote_ms < 0) {
        stderr_log("WARNING Invalid transmit timestamp in NTP response");
        return -1;
    }

    return (0, $local_before, $remote_ms, $local_after, $ip);
}

# Main execution
my $local_before_ms = 0;
my $remote_ms       = 0;
my $local_after_ms  = 0;
my $server_addr     = '';
my $success         = 0;

for my $attempt (1 .. $config{retries}) {
    $server_addr = '';
    if ($config{verbose}) {
        stderr_log("DEBUG Attempt ($attempt) at NTP query on $config{server} ...");
    }

    my ($rc, $before, $remote, $after, $ip) =
        do_ntp_query($config{server}, $config{timeout_ms});

    if (defined $rc && $rc == 0) {
        $local_before_ms = $before;
        $remote_ms       = $remote;
        $local_after_ms  = $after;
        $server_addr     = $ip;
        $success         = 1;
        last;
    }

    # small backoff before retry
    select(undef, undef, undef, 0.2);
}

unless ($success) {
    stderr_log("ERROR Failed to contact NTP server $config{server} after $config{retries} attempts");
    log_syslog(LOG_ERR, "NTP query failed for $config{server} after $config{retries} attempts");
    closelog() if $config{use_syslog};
    exit 2;
}

# Estimate network delay and clock offset like simple SNTP:
# t1 = client send time (local_before), t2 = server receive (not available),
# t3 = server transmit (remote_ms), t4 = client receive (local_after).
# offset = ((t3 + t2) - (t1 + t4)) / 2, but since t2 is unknown:
# offset ≈ t3 - ((t1 + t4) / 2)
# Check for potential overflow in avg calculation
if ($local_before_ms > 9223372036854775807 - $local_after_ms) {
    stderr_log("ERROR Time averaging would overflow, invalid timestamps.");
    log_syslog(LOG_ERR, "Time averaging would overflow");
    closelog() if $config{use_syslog};
    exit 1;
}
my $avg_local_ms  = int(($local_before_ms + $local_after_ms) / 2);
my $offset_ms     = $remote_ms - $avg_local_ms;
my $roundtrip_ms  = $local_after_ms - $local_before_ms;

# Parse remote time unconditionally — needed for year check and success log
my $remote_time_str = format_time_local($remote_ms);
my $remote_year     = (localtime(int($remote_ms / 1000)))[5] + 1900;

if ($config{verbose}) {
    my $local_time_str  = format_time_local($local_before_ms);
    my $local_after_str = sprintf("%03d", $local_after_ms % 1000);
    my $remote_ms_str   = sprintf("%03d", $remote_ms % 1000);

    stderr_log("DEBUG Server: $config{server} ($server_addr)");
    stderr_log("DEBUG Local time: $local_time_str.$local_after_str");
    stderr_log("DEBUG Remote time: $remote_time_str.$remote_ms_str");
    stderr_log("DEBUG Local before(ms): $local_before_ms");
    stderr_log("DEBUG Local after(ms): $local_after_ms");
    stderr_log("DEBUG Estimated roundtrip(ms): $roundtrip_ms");
    stderr_log("DEBUG Estimated offset remote - local(ms): $offset_ms");
    log_syslog(LOG_INFO,
        "NTP server=$config{server} addr=$server_addr offset_ms=$offset_ms rtt_ms=$roundtrip_ms");
}

# Basic sanity check on roundtrip
if ($roundtrip_ms < 0 || $roundtrip_ms > 10000) {
    stderr_log("ERROR Invalid roundtrip time: $roundtrip_ms ms");
    log_syslog(LOG_ERR, "Invalid suspiciously long roundtrip time: $roundtrip_ms ms");
    closelog() if $config{use_syslog};
    exit 1;
}

# Check if offset is within acceptable range (no adjustment needed)
my $abs_offset = abs($offset_ms);
if ($abs_offset > 0 && $abs_offset < 500) {
    if ($config{verbose}) {
        stderr_log("INFO Delta < 500ms, not setting system time.");
        log_syslog(LOG_INFO, "Delta < 500ms, not setting system time");
    }
    closelog() if $config{use_syslog};
    exit 0;
}

# Validate remote year
if ($remote_year < 2025 || $remote_year > 2200) {
    stderr_log("ERROR Remote year is out of valid range (2025-2200): $remote_year");
    log_syslog(LOG_ERR, "Remote year is out of valid range (2025-2200): $remote_year");
    closelog() if $config{use_syslog};
    exit 1;
}

# Test mode — return without adjusting
if ($config{test_only}) {
    closelog() if $config{use_syslog};
    exit 0;
}

# Check if running as root
if ($< != 0) {
    stderr_log("WARNING Not root, not setting system time.");
    log_syslog(LOG_WARNING, "Not root, not setting system time");
    closelog() if $config{use_syslog};
    exit 0;
}

# Set system time using settimeofday
my $half_rtt    = int($roundtrip_ms / 2);
# Check for potential overflow before time calculation
if ($remote_ms > 9223372036854775807 - $half_rtt) {
    stderr_log("ERROR Time calculation would overflow, not adjusting system time.");
    log_syslog(LOG_ERR, "Time calculation would overflow");
    closelog() if $config{use_syslog};
    exit 1;
}
my $new_time_ms   = $remote_ms + $half_rtt;
my $new_time_sec  = int($new_time_ms / 1000);
my $new_time_usec = ($new_time_ms % 1000) * 1000;

eval {
    require Time::HiRes;
    Time::HiRes->import('settimeofday');
};

if ($@) {
    stderr_log("ERROR Time::HiRes::settimeofday not available");
    log_syslog(LOG_ERR, "Time::HiRes::settimeofday not available");
    closelog() if $config{use_syslog};
    exit 10;
}

if (eval { Time::HiRes::settimeofday($new_time_sec, $new_time_usec); 1 }) {
    stderr_log(sprintf("INFO System time set using settimeofday (%s.%03d)",
        $remote_time_str, $remote_ms % 1000));
    log_syslog(LOG_INFO, sprintf("System time set using settimeofday (%s.%03d)",
        $remote_time_str, $remote_ms % 1000));
    closelog() if $config{use_syslog};
    exit 0;
} else {
    stderr_log("ERROR Failed to adjust system time with settimeofday: $!");
    log_syslog(LOG_ERR, "Failed to adjust system time with settimeofday: $!");
    closelog() if $config{use_syslog};
    exit 10;
}
